visual-js/visual/src/common/api.ts

```diff
@@ -148,30 +148,48 @@ const uploadToUrl = async ({
   contentType,
   file,
   compress,
+  uploadTimeoutMs,
 }: {
   uploadUrl: string;
   contentType: string;
   file: DataContent | DataPath;
   compress?: boolean;
+  uploadTimeoutMs?: number;
 }) => {
   const uploadBody = isDataPath(file) ? fs.readFileSync(file.path) : file.data;
 
   const hash = crypto.createHash('md5').update(uploadBody).digest('base64');
+  try {
+    uploadTimeoutMs ||= 10_000;
+    const result = await fetch(uploadUrl, {
+      method: 'PUT',
+      headers: {
+        'Content-Length': `${uploadBody.byteLength}`,
+        'Content-Type': contentType,
+        'Content-MD5': hash,
+      },
+      body: uploadBody,
+      compress,
+      ...fetchOptions,
+      signal: AbortSignal.timeout(uploadTimeoutMs),
+    });
 
-  const result = await fetch(uploadUrl, {
-    method: 'PUT',
-    headers: {
-      'Content-Length': `${uploadBody.byteLength}`,
-      'Content-Type': contentType,
-      'Content-MD5': hash,
-    },
-    body: uploadBody,
-    compress,
-    ...fetchOptions,
-  });
-
-  if (!result.ok) {
-    throw new Error(`Failed to upload snapshot: ${result.statusText}`);
+    if (!result.ok) {
+      throw new Error(`Failed to upload snapshot: ${result.statusText}`);
+    }
+  } catch (ex: unknown) {
+    const isObject = (value: unknown): value is any =>
+      !!value && typeof value === 'object';
+    if (isObject(ex)) {
+      if (ex?.name === 'AbortError') {
```

Comment:
```suggestion
    if (String(ex).includes('AbortError')) {
```

should be good enough and more readable